test/ngAnimate/animateRunnerSpec.js

```diff
@@ -0,0 +1,251 @@
+'use strict';
+
+describe("$$rAFMutex", function() {
+  beforeEach(module('ngAnimate'));
+
+  it('should fire the callback only when one or more RAFs have passed',
+    inject(function($$rAF, $$rAFMutex) {
+
+    var trigger = $$rAFMutex();
+    var called = false;
+    trigger(function() {
+      called = true;
+    });
+
+    expect(called).toBe(false);
+    $$rAF.flush();
+    expect(called).toBe(true);
+  }));
+
+  it('should immediately fire the callback if a RAF has passed since construction',
+    inject(function($$rAF, $$rAFMutex) {
+
+    var trigger = $$rAFMutex();
+    $$rAF.flush();
+
+    var called = false;
+    trigger(function() {
+      called = true;
+    });
+    expect(called).toBe(true);
+  }));
+});
+
+describe("$animateRunner", function() {
+
+  beforeEach(module('ngAnimate'));
+
+  they("should trigger the host $prop function",
+    ['end', 'cancel', 'pause', 'resume'], function(method) {
+
+    inject(function($animateRunner) {
+      var host = {};
+      var spy = host[method] = jasmine.createSpy();
+      var runner = new $animateRunner(host);
+      runner[method]();
+      expect(spy).toHaveBeenCalled();
+    });
+  });
+
+  they("should trigger the inner runner's host $prop function",
+    ['end', 'cancel', 'pause', 'resume'], function(method) {
+
+    inject(function($animateRunner) {
+      var host = {};
+      var spy = host[method] = jasmine.createSpy();
+      var runner1 = new $animateRunner();
+      var runner2 = new $animateRunner(host);
+      runner1.setHost(runner2);
+      runner1[method]();
+      expect(spy).toHaveBeenCalled();
+    });
+  });
+
+  it("should resolve the done function only if one RAF has passed",
+    inject(function($animateRunner, $$rAF) {
+
+    var runner = new $animateRunner();
+    var spy = jasmine.createSpy();
+    runner.done(spy);
+    runner.complete(true);
+    expect(spy).not.toHaveBeenCalled();
+    $$rAF.flush();
+    expect(spy).toHaveBeenCalled();
+  }));
+
+  it("should resolve with the status provided in the completion function",
+    inject(function($animateRunner, $$rAF) {
+
+    var runner = new $animateRunner();
+    var capturedValue;
+    runner.done(function(val) {
+      capturedValue = val;
+    });
+    runner.complete('special value');
+    $$rAF.flush();
+    expect(capturedValue).toBe('special value');
+  }));
+
+  they("should immediately resolve each combined runner in depth-first order when $prop is called",
```

Comment:
I don't think `depth-first` is an accurate description. You probably meant something like `bottom-up` or `fifo` or something.